Remove negativeMargin plugin, handle negative values in regular margin plugin

adamwathan opened the pull request:
The `negativeMargin` plugin has always sort of been an odd one out, and with #848 asking for the ability to mimic the `-mx-4`-style leading dash for negative values for other properties, it makes sense to solve this in a more general way.

This PR starts that process by eliminating the `negativeMargin` plugin, and making the regular `margin` plugin capable of handling negative values and generating the same `-mx-4`-style classes.

The way it works is if you specify a key in your `margin` configuration that starts with a `-`, the plugin translates that to a class name that starts with a `-`.

For example, this configuration:

```js
// tailwind.config.js
module.exports = {
  theme: {
    margin: {
      '-6': '-1.5rem',
    }
  }
}
```

...would generate these classes:

```css
.-m-6 { margin: -1.5rem }
.-my-6 { margin-top: -1.5rem; margin-bottom: -1.5rem }
.-mx-6 { margin-right: -1.5rem; margin-left: -1.5rem }
.-mt-6 { margin-top: -1.5rem }
.-mr-6 { margin-right: -1.5rem }
.-mb-6 { margin-bottom: -1.5rem }
.-ml-6 { margin-left: -1.5rem }
```

Importantly, **it does not generate classes like `.mx--6`** like you might expect, since normally keys in the config translate directly to class suffixes.

This is special/magic behavior, sort of like how the `default` key means "no suffix or trailing dash".

This approach was not decided on lightly, and it is unfortunate that we have to introduce new special behavior/magic, but it is the best solution in terms of balancing backwards compatibility, leveraging existing muscle memory, and class readability.

Alternatives that were considered included using classes like `mx-n6`, `mx-neg-6`, `mx--6`, `mx(-6)`, and even changing the whole framework to use colons instead of dashes so we could do `mx:-6`. Keeping the existing `-mx-6` class name and generalizing that approach so it works with other plugins is definitely the least bad option.

This is a breaking change in two ways:

1. **The negativeMargin key now does nothing in your config file**. Users will need to move any customizations made to this key to the `margin` key.

2. **Anyone already using keys like '-6' in their margin config will no longer see their existing classes generated**. Up until now, using a key like '-6' would result in classes like `mx--6`. Now `-mx-6` will be generated instead, and it is now impossible to generate values that result in a double dash in the class name. It looks stupid as hell anyways so honestly I'm doing you a favor.

The default `margin` config has been updated to automatically generate all of the negative values based on the `spacing` config:

```js
    // ...
    margin: theme => {
      const negativeSpacing = Object.keys(theme('spacing'))
        .filter(key => theme('spacing')[key] !== '0')
        .reduce((negativeSpacing, key) => {
          return { ...negativeSpacing, [`-${key}`]: `-${theme('spacing')[key]}` }
        }, {})

      return {
        auto: 'auto',
        ...theme('spacing'),
        ...negativeSpacing,
      }
    },
```

Having this logic live right in the config file is not my favorite thing in the world, especially because people can publish this with `tailwind init --full`, but it's fine for now and I can move it somewhere else after I think about the best place to put it.

## Next steps

The next step is to generalize this across all relevant plugins, starting with at least the `inset` and `zIndex` plugins.

I may also expose a new `className` function through the plugin API that combines this logic, the escaping logic, and the special `default` key handling into one function, so all plugin authors (including those working on Tailwind's core plugins) don't have to really worry about this stuff. It would make it really easy for everything to be consistent:

```js
function ({ addUtilities, className }) {
  const utilities = _.fromPairs(
    _.map(theme('rotate'), (value, modifier) => {
      return [
        `.${className('rotate', modifier)}`,
        {
          'transform': `rotate(${value})`,
        },
      ]
    })
  )

  addUtilities(utilities, variants('rotate'))
}
```

adamwathan committed:
Remove negativeMargin plugin, handle negative values in regular margin plugin

MichaelDeBoey commented on stubs/defaultConfig.stub.js:
```diff
@@ -305,8 +307,19 @@ module.exports = {
+      const negativeSpacing = Object.keys(theme('spacing'))
+        .filter(key => theme('spacing')[key] !== '0')
+        .reduce((negativeSpacing, key) => {
+          return { ...negativeSpacing, [`-${key}`]: `-${theme('spacing')[key]}` }
```
Maybe an implicit return would make the `reduce` a bit nicer?

```suggestion
.reduce((negativeSpacing, key) => ({
  ...negativeSpacing,
  [`-${key}`]: `-${theme('spacing')[key]}`
}))
```

adamwathan replied:
Yeah that looks a bit nicer, thanks!

MichaelDeBoey commented on stubs/defaultConfig.stub.js:
```diff
@@ -1,3 +1,5 @@
+const _ = require('lodash')
```
You're not using `lodash` in this file, so this import can be deleted, no?

adamwathan replied:
Yep good catch, thanks!

MichaelDeBoey commented on src/plugins/margin.js:
```diff
@@ -1,20 +1,30 @@
 import _ from 'lodash'
```
I think it's better to tree-shake the used `lodash` functions instead of importing the whole library

```suggestion
import flatMap from 'lodash/flatMap'
import startsWith from 'lodash/startsWith'
```

MichaelDeBoey replied:
If you want I can make a PR to do this in every file btw 🙂

adamwathan replied:
Since this runs in node and not the browser it doesn't matter honestly. No bundle size to worry about.

adamwathan committed:
Remove unused lodash dependency, use implicit return

adamwathan committed:
Move negative margin logic into a helper

Adds a new `utils` bucket that's passed as a second arg when using a closure for theme values. The idea is you can destructure useful helper functions out of this argument, in this case a `negative` function that converts a positive scale to negative values. That's the only helper function right now, but making it a destructurable arg so we can add more if necessary without adding a bunch of positional arguments.

john-griffin commented:
Just trying this out in our app, we had a case where we were using `md:-mt-0` which appears to have been removed now. We have found it useful to have negative margin 0 classes to cancel stuff out at certain responsive sizes.

benface commented:
@john-griffin A negative margin of 0 is the same as a margin of 0. You should be able to replace that class with `md:mt-0`.

john-griffin commented:
Thanks, just wanted to check that was intended way to move forward. Will probably need to be added to upgrade docs.
